src/librustc_mir/interpret/intrinsics/caller_location.rs

```diff
@@ -14,16 +14,19 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     /// Walks up the callstack from the intrinsic's callsite, searching for the first callsite in a
     /// frame which is not `#[track_caller]`. If the first frame found lacks `#[track_caller]`, then
     /// `None` is returned and the callsite of the function invocation itself should be used.
-    crate fn find_closest_untracked_caller_location(&self) -> Option<Span> {
-        let mut caller_span = None;
-        for next_caller in self.stack.iter().rev() {
-            if !next_caller.instance.def.requires_caller_location(*self.tcx) {
-                return caller_span;
-            }
-            caller_span = Some(next_caller.span);
-        }
-
-        caller_span
+    crate fn find_closest_untracked_caller_location(&self) -> Span {
+        self.stack
+            .iter()
+            .rev()
+            // Skip `#[track_caller]` frames.
+            .skip_while(|frame| frame.instance.def.requires_caller_location(*self.tcx))
+            // Find next frame with source info.
+            .find_map(|frame| frame.current_source_info())
```

Comment:
I'm noot sure this makes sense, if that frame doesn't have a source info, we'd now be returning the wrong call site, right? What does codegen do in this case?

cc @anp

Reply:
The only case where a frame does not have source info, is when its current block is `None`, which means that we are unwinding and the current frame has no cleanup code. In that case we'd skip to the caller, which seems like a reasonable choice to me.

I *think* this case is actually impossible (AFAIK only the top frame can ever have `None`, and then it will never call this intrinsic), so I'd also be fine just making it ICE to be proven wrong about that.

Reply:
Yes I prefer the ICE over silently doing the wrong thing in a situation we don't even know may exist